src/pipeline/events.rs

```diff
@@ -119,3 +122,110 @@ impl<'a> EventHandler for EventQueue<'a> {
         }
     }
 }
+
+#[cfg(test)]
+mod test {
+    use bevy::time::{TimePlugin, TimeUpdateStrategy};
+    use systems::tests::HeadlessRenderPlugin;
+
+    use crate::{plugin::*, prelude::*};
+
+    #[cfg(feature = "dim3")]
+    fn cuboid(hx: Real, hy: Real, hz: Real) -> Collider {
+        Collider::cuboid(hx, hy, hz)
+    }
+    #[cfg(feature = "dim2")]
+    fn cuboid(hx: Real, hy: Real, _hz: Real) -> Collider {
+        Collider::cuboid(hx, hy)
+    }
+
+    #[test]
+    pub fn events_received() {
+        return main();
+
+        use bevy::prelude::*;
+
+        #[derive(Resource, Reflect)]
+        pub struct EventsSaver<E: Event> {
+            pub events: Vec<E>,
+        }
+        impl<E: Event> Default for EventsSaver<E> {
+            fn default() -> Self {
+                Self {
+                    events: Default::default(),
+                }
+            }
+        }
+        pub fn save_events<E: Event + Clone>(
+            mut events: EventReader<E>,
+            mut saver: ResMut<EventsSaver<E>>,
+        ) {
+            for event in events.read() {
+                saver.events.push(event.clone());
+            }
+        }
+        fn run_test(app: &mut App) {
```

Comment:
Looks better with some spaces.
```suggestion
        #[derive(Resource, Reflect)]
        pub struct EventsSaver<E: Event> {
            pub events: Vec<E>,
        }

        impl<E: Event> Default for EventsSaver<E> {
            fn default() -> Self {
                Self {
                    events: Default::default(),
                }
            }
        }

        pub fn save_events<E: Event + Clone>(
            mut events: EventReader<E>,
            mut saver: ResMut<EventsSaver<E>>,
        ) {
            for event in events.read() {
                saver.events.push(event.clone());
            }
        }

        fn run_test(app: &mut App) {
```